File: 061.rb
# 061
# Rotate List
# 2015-12-16
#####################################################
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
  def initialize(val)
    @val = val
    @next = nil
  end
end

# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def rotate_right(head, k)
  return nil if not head
  
  tail, newtail, length = head, head, 1
  while tail.next
    length += 1
    tail = tail.next
  end
  tail.next = head

  k = length-1-k%length
  k.times { newtail = newtail.next }
  newhead = newtail.next
  newtail.next = nil
  
  newhead
end

#####################################################
a1 = ListNode.new(1)
a2 = ListNode.new(2)
a3 = ListNode.new(3)
a4 = ListNode.new(4)
a5 = ListNode.new(5)
a1.next = a2
a2.next = a3
a3.next = a4
a4.next = a5

k = 2

r = rotate_right(a1, k)
while r
  print r.val,' -> '
  r = r.next
end

File: 061.js
//  061
//  Rotate List
//  2015-12-16
/////////////////////////////////////////////////////
/**
 * Definition for singly-linked list.
 */
function ListNode(val) {
    this.val = val;
    this.next = null;
}
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head)
        return null;

    var tail = head, newtail = head, length = 1;
    while (tail.next) {
        length++;
        tail = tail.next;
    }
    tail.next = head
    
    k = length-1-k%length;
    for (var i = 0; i < k; i++) {
        newtail = newtail.next;
    }
    newhead = newtail.next;
    newtail.next = null;

    return newhead;
};

/////////////////////////////////////////////////////

var a1 = new ListNode(1);
var a2 = new ListNode(2);
var a3 = new ListNode(3);
var a4 = new ListNode(4);
var a5 = new ListNode(5);
a1.next = a2;
a2.next = a3;
a3.next = a4;
a4.next = a5;

var k = 2;
var r = rotateRight(a1, k);
while (r) {
    console.log(r.val);
    r = r.next;
}

